### PR TITLE
Fix QuTiP interaction with numpy 1.20

### LINKED ISSUE #1433: QuTiP incompatible with numpy 1.20 due to Qobj.__array__
QuTiP is currently incompatible with numpy 1.20.  Tests will fail to even collect with an error such as
```
AttributeError: 'numpy.ndarray' object has no attribute 'dag'
```
and a lot of functionality will break - anything that requires `Qobj.eigenstates()` for example.

This is because `Qobj` defines `__array__`, one of numpy's "array interface" functions, intended for classes that can be safely converted _implicitly_ into an `ndarray`.  This isn't really the case for `Qobj` - it loses all sorts of information when you do that, which is why we've maintained the separate `Qobj.full()` for explicitly getting the dense matrix representation of a `Qobj`.  This is not to mention that numpy ufuncs probably should not be able to implicitly convert `Qobj` - I'd strongly argue that `np.sin(qutip.basis(2, 1))` should be `TypeError`, _not_ `array([[0.        +0.j], [0.84147098+0.j]])` (like it is right now).  We actually already removed this "functionality" in `dev.major`.

The most pressing incompatibility is that a few points in QuTiP put a few `Qobj` into a `np.array(dtype=object)`.  In numpy 1.20, this no longer produces a 1D array of `Qobj`, but a 3D array of complex.  This breaks `Qobj.eigenstates`, and prevents test collection due to it being present in `states.py::qutrit_basis()`, which is called during parametrisation.

There are a two possible ways to solve this, and we ought to release a fix with one of them in a patch ASAP:
  1. remove `Qobj.__array__`
  2. remove all use of `Qobj` in `np.array`

I'm personally in favour of "explicit is better than implicit" in this case, i.e. removing `Qobj.__array__` and relying on `Qobj.full()`.  As another example along this vein, note `scipy.sparse` matrices don't implement this either, and they're arguably closer to being safely coerced to `ndarray` than we are.

### Related issues

#938: feature request for implementing `__array__`.  This only asks for `np.array(qobj)` as a convenience, acknowledging the availability of `Qobj.full()`.
#1017: includes a comment on buggy behaviour caused by `__array__`.  Note that the solution given there (`np.asarray(..., dtype=object)`) will no longer work with numpy 1.20.

### PR DESCRIPTION
**Pull request by jakelishman**
Fix #1433, see further discussion of points there.

- sets `Qobj.__array_ufunc__ = None` to prevent implicit usage with ufuncs
- change use of `np.array(dtype=object)` with `Qobj` into either raw Python lists (when only used internally), or sets `out = np.empty(shape, dtype=object); out[:] = ...` to force object-array creation
- fix version of numpy build version to `1.16.6 <= x < 1.20` to prevent cross-version ABI incompatibility (see also comments in #1429)
- change default of `dims` and `shape` in `Qobj` constructor to `None`, rather than dangerous lists (also helps avoid numpy calls on ragged sequences)

### CODE REVIEW

**Review by BoxiLi — COMMENTED** [review]
Looks good to me! Only two comments for clarification.

File: qutip/eseries.py [hunk 1]
```diff
@@ -84,18 +84,18 @@ def __init__(self, q=None, s=np.array([])):
...
-                ind = np.shape(q)
-                num = ind[0]  # number of elements in q
+                num = len(q)  # number of elements in q
                 if any([Qobj(x).shape != Qobj(q[0]).shape for x in q]):
                     raise TypeError('All amplitudes must have same dimension.')
-                self.ampl = np.array([x for x in q], dtype=object)
-                self.rates = np.zeros(ind)
+                self.ampl = np.empty((num,), dtype=object)
+                self.ampl[:] = q
+                self.rates = np.zeros((num,))
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply qutip/eseries.py:92]: dunno - habit, I guess?  Technically the first argument to most numpy constructors is a shape (if it's scalar, it gets promoted to a 1-tuple), so I guess that's why I do it?  There's no difference between the two forms.

File: qutip/qobj.py [hunk 2]
```diff
@@ -1698,15 +1722,20 @@ def eigenstates(self, sparse=False, sort='low', eigvals=0,
...
                                maxiter=maxiter)
-        new_dims = [self.dims[0], [1] * len(self.dims[0])]
-        ekets = np.array([Qobj(vec, dims=new_dims) for vec in evecs],
-                         dtype=object)
+        if self.type == 'super':
+            new_dims = [self.dims[0], [1]]
+            new_type = 'operator-ket'
+        else:
+            new_dims = [self.dims[0], [1] * len(self.dims[0])]
+            new_type = 'ket'
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply qutip/qobj.py:1730]: Oh, it's kind of a mistake this got caught in this patch - it's something I spotted while fixing the numpy errors in `essolve` and before I noticed that `essolve` was returning objects with the dimensions set wrong, I thought it was the cause of numpy errors coming from `Qobj.__init__`.

So I thought it was related to the numpy errors, but really it was a bug that happened to be spotted while fixing by them.  I can rebase it out of this PR and into a different one.

File: qutip/eseries.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jakelishman) [reply qutip/eseries.py:92]: Maybe I was making a like-for-like replacement - in the original code, `ind` was the output of `np.shape`, so should have been a 1-tuple.  I don't really know - I sort of just write numpy constructors on autopilot!

**Review by BoxiLi — APPROVED** [review]

**Pull request merged**